### PR TITLE
[presets] Add speed preset ⚡️

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
Adds torch compile support for LG2 (6x speed-ups), including fusing the q/k/v proj when compiled (6.7x vs baseline).

TODO: 
- [ ] Add torch export for saving compiled graphs

### CODE REVIEW

File: local_gemma/attention.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        bsz, q_len, _ = hidden_states.size()
+
+        query_key_value_states = self.qkv_proj(hidden_states)
+        query_size = self.num_heads * self.head_dim
+        key_value_size = self.num_key_value_heads * self.head_dim
+        query_states, key_states, value_states = query_key_value_states.split(
+            [query_size, key_value_size, key_value_size], dim=-1
+        )
+
+        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
```
Comment (sanchit-gandhi) [inline local_gemma/attention.py:93]: The remainder of this attention class is one-to-one the same as that in Transformers: https://github.com/huggingface/transformers/blob/535fe78b9f1d148684723e51f00645351880c47a/src/transformers/models/gemma2/modeling_gemma2.py#L224

We'd ideally have a mechanism to keep the attention modules sync'd as a follow-up PR

File: local_gemma/cli.py [hunk 2]
```diff
@@ -194,18 +209,40 @@ def main():
...
+            # when generating using max_new_tokens, update the cache on each generation step to limit memory
+            cache = None
 
         if device == "mps" and args.preset == "auto" and args.max_new_tokens is None:
             print("Setting max new tokens to 1024 for faster mps generation. To bypass this limit, set `--max_new_tokens=2048`.")
             args.max_new_tokens = 1024
+        elif (device == "cuda" and args.preset == "speed") and (not has_starting_prompt or is_instruction_tuned):
+            print("Compiling the model forward pass. This may take a few minutes, particularly the first time it is run...")
+            padding = "longest"
+            pad_to_multiple_of = model.config.max_position_embeddings // 4
```
Comment (sanchit-gandhi) [inline local_gemma/cli.py:234]: Currently pre-compiling 3 different cache lengths

File: local_gemma/modeling_local_gemma_2.py [hunk 3]
```diff
@@ -169,4 +186,23 @@ def from_pretrained(
...
             model.to(device, dtype=preset_kwargs["torch_dtype"])
 
+        if torch_compile and device != "cuda":
+            raise ValueError(
+                "Torch compile is only compatible with cuda devices. Set `torch_compile=False` in `.from_pretrained`"
+                f"for device {device}."
+            )
+        elif torch_compile:
+            model = fuse_attention_weights(model, device, torch_dtype)
+            model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)
```
Comment (sanchit-gandhi) [inline local_gemma/modeling_local_gemma_2.py:196]: Would be curious to hear whether you agree with this design @gante. I've stayed consistent with our current `from_pretrained` method which handles everything for the user (e.g. placing the model on the correct the `device`). However, it goes a bit against PyTorch design (which would return an `nn.Module` to the user and have them apply compile themselves).

Comment (gante) [reply local_gemma/modeling_local_gemma_2.py:196]: I'd say that, for the purposes of repo, this is positive! 👍

**Comment by gante** [comment]
@sanchit-gandhi possibly some recompilation is happening in the CLI -- perhaps the cache is being re-initialized due to different maximum lengths?

torch.export (see convo on slack) should also help!

**Comment by sanchit-gandhi** [comment]
No recompilation according to `torch._logging.set_logs(graph_breaks=True, recompiles=True)`, but terrible GPU utilisation on the torch profile:
Screenshot 2024-07-29 at 22 19 04

Wondering if the CLI is interfering with the inductor - going to dig into this more

File: local_gemma/cli.py [hunk 4]
```diff
@@ -194,18 +217,40 @@ def main():
...
+                dtype=model.dtype,
+            )
+            model.generation_config.cache_implementation = None
+        else:
+            # when generating using max_new_tokens, update the cache on each generation step to limit memory
+            cache = None
+
+        if hasattr(model.forward, "_torchdynamo_orig_callable"):
+            print("Compiling the model forward pass. This may take a few minutes, particularly the first time it is run...")
+            for _ in range(2):
```
Comment (sanchit-gandhi) [inline local_gemma/cli.py:239]: Temporary fix for https://github.com/huggingface/transformers/issues/32313

File: local_gemma/cli.py [hunk 5]
```diff
@@ -11,19 +11,25 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import os
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
+os.environ["TORCHINDUCTOR_FX_GRAPH_CACHE"] = "1"
```
Comment (sanchit-gandhi) [inline local_gemma/cli.py:16]: https://pytorch.org/tutorials/recipes/torch_compile_caching_tutorial.html

**Comment by sanchit-gandhi** [comment]
Merging this to unblock assisted generation PR @gante!

**Pull request merged**